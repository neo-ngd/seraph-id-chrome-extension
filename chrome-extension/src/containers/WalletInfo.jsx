// Copyright (c) 2019 Swisscom Blockchain AG
// Licensed under MIT License

import React, { useState, useEffect } from 'react';
import { Box } from '@material-ui/core';
import Claim from '../components/Cards/Claim';
import { decrypt } from '../commons/seraphSdkUtils';
import { useDispatch, useSelector } from 'react-redux';
import NavBar from '../components/NavBar/NavBar';
import Layout from '../components/Layout/Layout';
import AccountsModal from '../components/Modals/AccountsModal';
import { setExportedWallet } from '../pages/Background/actions';
import { getEncryptedPassword } from '../pages/Background/actions';
import { ENCRYPTED_PW_MSG } from '../commons/constants';
import dictionary from '../commons/dictionary';
import env from '../environments/environment';

/**
 * <WalletInfo />
 * Base view which is shown when the user is logged in.
 * @return {*}
 * @constructor
 */
const WalletInfo = () => {
  const dispatch = useDispatch();
  const [wallet, setWallet] = useState(null);
  const [pw, setPw] = useState(null);
  const [modalVisibility, setModalVisibility] = useState(false);
  const [isLoading, setIsLoading] = useState();
  const { activeAccount } = useSelector((state) => state);

  useEffect(() => {
    setIsLoading(true);
    chrome.runtime.onMessage.addListener(listener);
    dispatch(getEncryptedPassword());
    return () => chrome.runtime.onMessage.removeListener(listener);
  }, []);

  /**
   * The listener which decrypt the wallet using the provided data.
   * @param request
   * @return {Promise<void>}
   */
  const listener = async (request) => {
    if (request.msg === ENCRYPTED_PW_MSG) {
      setPw(request.password);
      await decryptAccount(request.password, request.wallet);
    }
  };

  /**
   * Decrypt the wallet
   * @param password
   * @param accountFromStore
   * @return {Promise<void>}
   */
  const decryptAccount = async (password, accountFromStore) => {
    const wallet = await decrypt(accountFromStore, password);
    if (!!wallet) {
      setWallet(wallet);
      setIsLoading(false);
    }
  };

  /**
   * Remove the claim for the wallet.
   * @param id
   * @return {Promise<void>}
   */
  const removeClaim = async (id) => {
    delete wallet.accounts[0].claims[id];
    setWallet(wallet);

    await wallet.accounts[0].encrypt(pw);
    const exportedWalletJSON = JSON.stringify(wallet.export());
    dispatch(setExportedWallet(exportedWalletJSON));
  };

  /**
   * Return claim list.
   * @param claims
   * @return {*}
   */
  const showAllClaims = (claims) => {
    return claims.map((claim) => (
      <Claim
        key={claim[0]}
        id={claim[1].id}
        schema={claim[1].schema}
        content={claim[1].attributes}
        onRemoveClaim={removeClaim}
      />
    ));
  };

  /**
   * Open accounts modal.
   */
  const openAccountsModal = () => {
    setModalVisibility(true);
  };

  /**
   * Close accounts modal
   */
  const handleCloseAccountsModal = () => {
    setModalVisibility(false);
  };

  /**
   * Open the new browser tab with the external demo site.
   */
  const openDemo = () => {
    chrome.tabs.create({ url: env.DEMO_URL });
  };

  if (!!wallet) {
    const account = wallet.accounts.find((acc) => acc.label === activeAccount);
    const claimsArr = Object.entries(account.claims);
    const { label: address } = account;

    return (
      <Layout padding={'60px 0 0 0'} justifyStart isLoading={isLoading}>
        <NavBar
          address={address}
          onOpenAccountsModal={openAccountsModal}
          name={`${dictionary.commons.account} ${wallet.accounts.indexOf(
            account
          ) + 1}`}
        />
        <Box
          display="flex"
          flex="1"
          flexDirection="column"
          justifyContent={claimsArr.length > 0 ? 'flex-start' : 'space-between'}
          overflow="auto"
        >
          <Box fontSize="24px" color="text.primary">
            {dictionary.commons.claims}
          </Box>
          {claimsArr.length > 0 ? (
            <Box>{showAllClaims(claimsArr)}</Box>
          ) : (
            <Box
              color="text.primary"
              fontSize="16px"
              lineHeight="28px"
              textAlign="center"
            >
              {dictionary.walletInfo.info}
              <span
                data-test-id={'open-demo-link'}
                style={{ cursor: 'pointer', textDecoration: 'underline' }}
                onClick={openDemo}
              >
                {dictionary.walletInfo.website}
              </span>
            </Box>
          )}
        </Box>

        <AccountsModal
          onClose={handleCloseAccountsModal}
          isOpen={modalVisibility}
          wallet={wallet}
        />
      </Layout>
    );
  }
  return <Layout isLoading={isLoading} />;
};

export default WalletInfo;
