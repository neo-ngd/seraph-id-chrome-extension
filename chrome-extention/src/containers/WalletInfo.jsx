import React, { useState, useEffect } from 'react';
import { Box } from '@material-ui/core';
import Claim from '../components/Cards/Claim';
import { createWallet } from '../commons/seraphSdkUtils';
import NavBar from '../components/NavBar/NavBar';
import Layout from '../components/Layout/Layout';
import AccountsModal from '../components/Modals/AccountsModal';
import { useDispatch } from "react-redux";
import {getEncryptedPassword} from "../pages/Background/actions";

function WalletInfo({ accountFromStore }) {
  const dispatch = useDispatch();
  const [wallet, setWallet] = useState(null);
  const [modalVisibility, setModalVisibility] = useState(false);

  useEffect(() => {
    chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
      if (request.msg === 'encryptedPw') {
        decryptAccount(request.password);
      }
    });
    dispatch(getEncryptedPassword())
  }, []);

  const decryptAccount = async (password) => {
    const importedWallet = createWallet(JSON.parse(accountFromStore));
    await importedWallet.accounts[0].decrypt(password);
    setWallet(importedWallet);
  };

  
  const showAllClaims = () => {
    const claimsArr = Object.entries(wallet.accounts[0].claims);
    return claimsArr.map((claim) => (
      <Claim key={claim[0]} id={claim[1].id} schema={claim[1].schema} content={claim[1].attributes} />
    ));
  };

  const openAccountsModal = () => {
    setModalVisibility(true);
  };

  const handleCloseAccountsModal = () => {
    setModalVisibility(false);
  };


  if (wallet) {
    const { label: address } = wallet.accounts[0];
    const claimsArr = Object.entries(wallet.accounts[0].claims);

    return (
      <Layout padding={'60px 0 0 0'} justifyStart>
        <NavBar address={address} onOpenAccountsModal={openAccountsModal} name="Account 1" />
        <Box
          display="flex" 
          flex="1"
          flexDirection="column"
          justifyContent={claimsArr.length > 0 ? 'flex-start' : 'space-between'}
        >
          <Box fontSize="24px" color="text.primary">Claims</Box>

          {claimsArr.length > 0 ? (
            <Box>
              {showAllClaims()}
            </Box>
          ) : (
            <Box color="text.primary" fontSize="16px" lineHeight="28px" textAlign="center">
              No Claims yet? Why you donâ€™t play around with our demo
            </Box>
          )}
        </Box>

        <AccountsModal
          onClose={handleCloseAccountsModal}
          isOpen={modalVisibility}
          wallet={wallet}
        />
      </Layout>
    );
  }
  return null;
}

export default WalletInfo;
