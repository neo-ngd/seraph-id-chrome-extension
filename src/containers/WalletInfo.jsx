import React, { useState, useEffect } from 'react';
import Grid from '@material-ui/core/Grid';
import Divider from '@material-ui/core/Divider';
import Typography from '@material-ui/core/Typography';
import CopyButton from '../components/Buttons/CopyButton';
import Claim from '../components/Cards/Claim';
import { createWallet } from '../commons/seraphSdkUtils';
import { useSelector } from 'react-redux';

function WalletInfo({ accountFromStore }) {
  const [wallet, setWallet] = useState(null);
  const password = useSelector((state) => state.password);
  useEffect(() => {
    async function decryptAccount() {
      const importedWallet = createWallet(JSON.parse(accountFromStore));
      await importedWallet.accounts[0].decrypt(password);
      setWallet(importedWallet);
    }
    decryptAccount();
  }, []);

  function showAllClaims() {
    var claimsArr = Object.entries(wallet.accounts[0].claims);
    let list = claimsArr.map((claim) => (
      <Claim id={claim[1].id} schema={claim[1].schema}></Claim>
    ));
    return list;
  }

  if (wallet) {
    const { label: address } = wallet.accounts[0];
    return (
      <React.Fragment>
        <CopyButton textToCopy={address}>
          <Grid container>
            <Grid item xs={12}>
              <Typography variant="button">{wallet.name}</Typography>
            </Grid>
            <Grid item xs={12}>
              <Typography variant="overline" display="block" gutterBottom>
                {address}
              </Typography>
            </Grid>
          </Grid>
        </CopyButton>
        <Divider />
        <Typography variant="h6" display="block" gutterBottom>
          Claims
        </Typography>
        {showAllClaims()}
      </React.Fragment>
    );
  } else return null;
}

export default WalletInfo;
