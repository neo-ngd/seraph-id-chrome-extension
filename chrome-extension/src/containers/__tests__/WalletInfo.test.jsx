import React from 'react';
import { mount } from 'enzyme';
import WalletInfo from "../WalletInfo";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import chrome from 'sinon-chrome';

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('<WalletInfo />', () => {
    const store = mockStore({});
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, setState]);

    beforeEach(() => {
        global.chrome = chrome;
    });

    afterEach(() => {
        jest.clearAllMocks();
        store.clearActions();
    });

    it('should render component', () => {
        const wrapper = mount(
            <Provider store={store}>
                <WalletInfo />
            </Provider>);
        expect(wrapper).toMatchSnapshot()
    });

    // it('should open demo page', () => {
    //     const wrapper = mount(
    //         <Provider store={store}>
    //             <WalletInfo  />
    //         </Provider>);
    //     wrapper.find({'data-test-id': 'open-demo-link'}).first().simulate('click');
    //     expect(chrome.tabs.create.calledOnce).toBeTruthy();
    // });
});
