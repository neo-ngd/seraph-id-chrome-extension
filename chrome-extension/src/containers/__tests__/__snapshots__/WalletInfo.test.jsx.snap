// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WalletInfo /> should render component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <WalletInfo>
    <Layout
      isLoading={true}
    >
      <Styled(MuiBox)
        display="flex"
        flex="0 0 100%"
        flexDirection="column"
        height="100%"
        justifyContent="space-between"
        maxWidth="100%"
        overflow="hidden"
      >
        <div
          className="MuiBox-root MuiBox-root-2"
        >
          <Styled(MuiBox)
            style={
              Object {
                "margin": "auto",
              }
            }
          >
            <div
              className="MuiBox-root MuiBox-root-3"
              style={
                Object {
                  "margin": "auto",
                }
              }
            >
              <WithStyles(ForwardRef(CircularProgress))
                style={
                  Object {
                    "color": "white",
                  }
                }
              >
                <ForwardRef(CircularProgress)
                  classes={
                    Object {
                      "circle": "MuiCircularProgress-circle",
                      "circleDisableShrink": "MuiCircularProgress-circleDisableShrink",
                      "circleIndeterminate": "MuiCircularProgress-circleIndeterminate",
                      "circleStatic": "MuiCircularProgress-circleStatic",
                      "colorPrimary": "MuiCircularProgress-colorPrimary",
                      "colorSecondary": "MuiCircularProgress-colorSecondary",
                      "indeterminate": "MuiCircularProgress-indeterminate",
                      "root": "MuiCircularProgress-root",
                      "static": "MuiCircularProgress-static",
                      "svg": "MuiCircularProgress-svg",
                    }
                  }
                  style={
                    Object {
                      "color": "white",
                    }
                  }
                >
                  <div
                    className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                    role="progressbar"
                    style={
                      Object {
                        "color": "white",
                        "height": 40,
                        "width": 40,
                      }
                    }
                  >
                    <svg
                      className="MuiCircularProgress-svg"
                      viewBox="22 22 44 44"
                    >
                      <circle
                        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                        cx={44}
                        cy={44}
                        fill="none"
                        r={20.2}
                        strokeWidth={3.6}
                        style={Object {}}
                      />
                    </svg>
                  </div>
                </ForwardRef(CircularProgress)>
              </WithStyles(ForwardRef(CircularProgress))>
            </div>
          </Styled(MuiBox)>
        </div>
      </Styled(MuiBox)>
    </Layout>
  </WalletInfo>
</Provider>
`;
